### PR TITLE
Fix the limit clause applying on the SegmentGeneral/General locus

### PR DESCRIPTION
**Pull request by gfphoenix78**
When creating a limit path on the top of a subpath, which is
unorder and with SegmentGeneral/General locus, the locus type
of the Limit path should be no longer SegmentGeneral/General.
Because it may produce different output on different segments.
For example:
```
gpadmin=# explain insert into rpt select i from rpt2 limit 1;
                                  QUERY PLAN
------------------------------------------------------------------------------
 Insert on rpt  (cost=0.00..431.03 rows=1 width=4)
   ->  Result  (cost=0.00..431.00 rows=1 width=8)
         ->  Limit  (cost=0.00..431.00 rows=1 width=4)
               ->  Seq Scan on rpt rpt2  (cost=0.00..431.00 rows=10 width=4)
```
The table `rpt` and `rpt2` are replicated tables. The subquery
`select i from rpt2 limit 1;` may be different on different segments,
which will lead the replicated table to an inconsistent state.

With this patch, the subquery is executed only on one segment, and
the result is broadcasted to all segments. The plan is:
```
gpadmin=# explain insert into rpt select i from rpt2 limit 1;
                                        QUERY PLAN
-------------------------------------------------------------------------------------------
 Insert on rpt  (cost=1063.00..1063.05 rows=1 width=4)
   ->  Broadcast Motion 1:3  (slice1; segments: 1)  (cost=1063.00..1063.05 rows=3 width=4)
         ->  Limit  (cost=1063.00..1063.00 rows=1 width=4)
               ->  Seq Scan on rpt2  (cost=0.00..1063.00 rows=96300 width=4)
```

Co-authored-by: Zhenghua Lyu zlv@pivotal.io

## Here are some reminders before you submit the pull request
- [ ] Add tests for the change
- [ ] Document changes
- [ ] Communicate in the mailing list if needed
- [ ] Pass `make installcheck`
- [ ] Review a PR in return to support the community

### CODE REVIEW

**Comment by gfphoenix78** [comment]
This PR is only done for PG planner, needs to be fixed for GPORCA.

**Comment by gfphoenix78** [comment]
@vraghavan78 This bug still exists in ORCA. Could you take some time to review this PR and fix it in ORCA?

File: src/backend/optimizer/util/pathnode.c [hunk 1]
```diff
@@ -4918,6 +4918,22 @@ create_limit_path(PlannerInfo *root, RelOptInfo *rel,
...
+	/*
+	 * If the locus type of the subpath is SegmentGeneral/General without order,
+	 * e.g. for replicated table, applying Limit will change the locus type.
+	 * Because the output of the subquery may be different, even for
+	 * replicated table, the result should be produced by one and only one
+	 * instance.
+	 */
+	if ((CdbPathLocus_IsGeneral(subpath->locus) ||
+		 CdbPathLocus_IsSegmentGeneral(subpath->locus))
+		&& !subpath->pathkeys)
```
Comment (hlinnaka) [inline src/backend/optimizer/util/pathnode.c:4930]: Even if there's *some* order, the result is nevertheless not deterministic. Consider `insert into rpt select i from rpt2 ORDER BY col1 limit 1`, the order is still not well defined for rows with the same `col1` value.

Comment (gfphoenix78) [reply src/backend/optimizer/util/pathnode.c:4930]: Got it. Thanks.

**Comment by vraghavan78** [comment]
@asubramanya can you take a look at this and see what needs to happen for Orca to fix the issue?

**Comment by asubramanya** [comment]
@vraghavan78 Looked into this. We will probably need to fix the distribution request for the `CPhysicalLimit` when its on top of replicated tables. I have created a story in the backlog to track this issue for now.

**Comment by gfphoenix78** [comment]
Some thoughts about the optimization:
Assume the subquery contains no volatile function.
1. If the list of order-by fields is unique in the replicated table, the order of the subquery is well defined.
2. If case 1 is no true, but the used columns in the target list from the replicated table are the subset of the set of order-by fields, the order of the subquery is also well defined. Because the subset of the order-by list is always ordered.
3. If we uses the index scan that well defines the order of the superset of the columns used in target list, the order of the subquery is also defined.

In the above 3 cases, we could safely keep the locus type(General/SegmentGeneral) unchanged.

File: src/test/regress/sql/limit_gp.sql [hunk 2]
```diff
@@ -66,3 +66,26 @@ select array(select b from t_limit_all order by b asc limit all) t;
...
 
 drop table t_limit_all;
+
+-- Check LIMIT on a subquery with General/SegmentGeneral locus
+create table t_rpt(i int) distributed replicated;
+create table t_rpt2(i int) distributed replicated;
+create table t_prt(i int);
+create table t_prt2(i int);
+
+set optimizer = off;
```
Comment (zhangh43) [inline src/test/regress/sql/limit_gp.sql:76]: we should at least comment here, when orca support it, `set optimizer = off` should be removed, right?

**Comment by gfphoenix78** [comment]
> Except the ORCA test, others looks good to me.

@zhangh43  The test case  for ORCA is fake, just for the pipeline happy. It will be updated by the ORCA fix.

File: src/test/regress/sql/limit_gp.sql (hunk 2 above)
Comment (gfphoenix78) [reply src/test/regress/sql/limit_gp.sql:76]: Sounds good.

**Comment by gfphoenix78** [comment]
@vraghavan78 @asubramanya Do you have any update with this issue for ORCA? I'll merge this PR. When you fixed this issue, please update the test case for ORCA. Thanks.

File: src/backend/optimizer/util/pathnode.c [hunk 3]
```diff
@@ -4923,6 +4923,27 @@ create_limit_path(PlannerInfo *root, RelOptInfo *rel,
...
+	 */
+	if ((CdbPathLocus_IsGeneral(subpath->locus) ||
+		 CdbPathLocus_IsSegmentGeneral(subpath->locus)))
+	{
+		CdbPathLocus locus;
+		CdbPathLocus_MakeSingleQE(&locus, subpath->locus.numsegments);
+		if (CdbPathLocus_IsGeneral(subpath->locus))
+			locus.numsegments = getgpsegmentCount();
+		Assert(locus.numsegments > 0);
+		subpath = cdbpath_create_motion_path(root, subpath, NIL, false, locus);
```
Comment (kainwen) [inline src/backend/optimizer/util/pathnode.c:4944]: Why pathkey parameter of `cdbpath_create_motion_path` is NIL?

Comment (JunfengYang) [reply src/backend/optimizer/util/pathnode.c:4944]: I'm also curious about why adding a motion path here? If the locus is `SingleQE`, the motion seems will get removed.

Comment (JunfengYang) [reply src/backend/optimizer/util/pathnode.c:4944]: As `create_limit_path` didn't add any motions before.

Comment (kainwen) [reply src/backend/optimizer/util/pathnode.c:4944]: > I'm also curious about why adding a motion path under the limit path? If the locus is `SingleQE`, the motion seems will get removed.

@JunfengYang  Sometimes segmentgeneral -> singleQE's motion cannot be removed.

Comment (JunfengYang) [reply src/backend/optimizer/util/pathnode.c:4944]: @kainwen Thanks! I was failed to build one, any examples?

Comment (kainwen) [reply src/backend/optimizer/util/pathnode.c:4944]: > @kainwen Thanks! I was failed to build one, any examples?

select * from pg_class, (select * from replicate_table limit 10); -- you cannot change the subquery's locus to singleQE directly.

Comment (JunfengYang) [reply src/backend/optimizer/util/pathnode.c:4944]: @kainwen understand, thanks a lot! Seems the tests not cover this case.

Comment (JunfengYang) [reply src/backend/optimizer/util/pathnode.c:4944]: For this plan
```
                                               QUERY PLAN
--------------------------------------------------------------------------------------------------------
 Hash Join  (cost=1063.22..1075.08 rows=11 width=231)
   Hash Cond: (pg_class.oid = (t_rpt.i)::oid)
   ->  Seq Scan on pg_class  (cost=0.00..10.59 rows=459 width=231)
   ->  Hash  (cost=1063.10..1063.10 rows=10 width=4)
         ->  Limit  (cost=1063.00..1063.00 rows=10 width=4)
               ->  Gather Motion 1:1  (slice1; segments: 1)  (cost=1063.00..1063.00 rows=96300 width=4)
                     ->  Seq Scan on t_rpt  (cost=0.00..1063.00 rows=96300 width=4)
 Slice 0: Dispatcher; root 0; parent -1; gang size 0
 Slice 1: Singleton Reader; root 0; parent 0; gang size 1; segment 2
```
I think the `LIMIT` should push under `MOTION` node to make the QE stop
early. But it's another story.

**Comment by asubramanya** [comment]
@gfphoenix78 https://github.com/greenplum-db/gpdb/pull/10380 has been opened to address the issue on ORCA side. I will update the test cases once I merge the ORCA PR.

File: src/backend/optimizer/util/pathnode.c (hunk 3 above)
Comment (gfphoenix78) [reply src/backend/optimizer/util/pathnode.c]: > Why pathkey parameter of `cdbpath_create_motion_path` is NIL?

Fixed. Thank you.

**Comment by kainwen** [comment]
A pr to solve a wider issue is opened: https://github.com/greenplum-db/gpdb/pull/10418
Related thread is: https://groups.google.com/a/greenplum.org/g/gpdb-dev/c/354vF_0EPIM

File: src/backend/optimizer/util/pathnode.c (hunk 3 above)
Comment (kainwen) [reply src/backend/optimizer/util/pathnode.c]: > I think the `LIMIT` should push under `MOTION` node to make the QE stop
> early. But it's another story.

@JunfengYang 
Good catch! 

A pr to solve a wider issue is opened: https://github.com/greenplum-db/gpdb/pull/10418
Related thread is: https://groups.google.com/a/greenplum.org/g/gpdb-dev/c/354vF_0EPIM

The new pr adopts your idea.

**Comment by gfphoenix78** [comment]
Close this PR in favor of #10418

**Pull request closed without merging**